src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -1666,8 +1761,21 @@ public async Task WebApplicationBuilder_OnlyAddsDefaultServicesOnce(CreateBuilde
         Assert.Single(app.Services.GetRequiredService<IEnumerable<IServer>>());
     }
 
+    [Fact]
+    public void EmptyWebApplicationBuilder_OnlyContainsMinimalServices()
+    {
+        var builder = WebApplication.CreateEmptyBuilder(new());
+
+        Assert.Empty(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IConfigureOptions<LoggerFactoryOptions>)));
+        Assert.Empty(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IServer)));
+
+        // These services are still necessary
+        Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IWebHostEnvironment)));
+        Assert.Single(builder.Services.Where(descriptor => descriptor.ServiceType == typeof(IOptionsChangeTokenSource<HostFilteringOptions>)));
```

Comment:
Why is this part of the empty builder?

Reply:
It comes from this code

https://github.com/dotnet/aspnetcore/blob/bd9f5d1df2137935bec66b711c7c1534978a421c/src/DefaultBuilder/src/WebHost.cs#L265-L268

from my understanding, there is no other public way to get host filtering enabled. The class is internal and no other code references it.

Reply:
I remember @blowdart saying that HostFiltering is recommended. See https://github.com/dotnet/aspnetcore/issues/47862.

Reply:
We don’t want another empty builder . This has to be the last low level builder API so it really needs to be empty

Reply:
Fixed.